alf/algorithms/algorithm.py

```diff
@@ -1050,11 +1081,28 @@ def update_with_gradient(self,
             assert len(loss_info.scalar_loss.shape) == 0
             loss_info = loss_info._replace(
                 loss=add_ignore_empty(loss_info.loss, loss_info.scalar_loss))
+        return loss_info
 
+    def _backward_and_gradient_update(
+            self,
+            loss_info,
+            weight=1.0,
+    ):
+        """Do backward and gradient update to all the trainable parameters.
+
+        Args:
+            loss_info (LossInfo): loss with shape :math:`(T, B)` (except for
+                ``loss_info.scalar_loss``)
+            weight (float): weight for this batch. Loss will be multiplied with
```

Comment:
This can be omitted if using a scalar for loss because it can be multiplied with the scalar loss by the caller.

Reply:
Now the weight is moved out side of the function and multiplied to the loss before calling this function.